File: backend/app/routes/texts.py
"""Text APIs; all file mutations are version checked by the service."""
from pathlib import Path
from functools import wraps

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import FileResponse
from pydantic import BaseModel, Field

from .. import texts

router = APIRouter(prefix='/api/texts', tags=['texts'])


def errors(fn):
    @wraps(fn)
    def call(*args, **kwargs):
        try:
            return fn(*args, **kwargs)
        except FileNotFoundError as e:
            raise HTTPException(404, '文件或目录不存在，编辑内容仍可另存副本') from e
        except (OSError, ValueError, UnicodeError) as e:
            raise HTTPException(400, str(e)) from e
    return call


class Paths(BaseModel):
    paths: list[str] = Field(min_length=1, max_length=100)


class Reorder(BaseModel):
    ids: list[int] = Field(min_length=2, max_length=10000)


@router.post('/reorder')
@errors
def reorder(payload: Reorder):
    return texts.reorder(payload.ids)


class Create(BaseModel):
    directory: str = ''
    folder_id: int | None = None
    name: str
    body: str = Field(default='', max_length=texts.MAX_BYTES)


class Save(BaseModel):
    body: str = Field(max_length=texts.MAX_BYTES)
    version: str
    encoding: str | None = None
    resolving_conflict: bool = False


class Meta(BaseModel):
    favorite: bool | None = None
    tags: list[str] | None = Field(default=None, max_length=100)


class Version(BaseModel):
    version: str


class Move(Version):
    directory: str = ''
    name: str = ''


@router.get('')
@errors
def listing(q: str = '', folder_id: int | None = None, view: str = 'all', format: str = '',
            tag: str = '', sort: str = 'mtime', offset: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=500)):
    return texts.listing(q, folder_id, view, format, tag, sort, offset, limit)


@router.post('/preview')
@errors
def preview(payload: Paths):
    return texts.preview(payload.paths)


@router.post('/associate')
@errors
def associate(payload: Paths):
    return texts.associate(payload.paths)


@router.post('')
@errors
def create(payload: Create):
    return texts.create(payload.directory, payload.name, payload.body, payload.folder_id)


@router.get('/{text_id}')
@errors
def detail(text_id: int, encoding: str | None = None, opened: bool = False):
    return texts.detail(text_id, encoding, opened)


@router.put('/{text_id}')
@errors
def save(text_id: int, payload: Save):
    return texts.save(text_id, payload.body, payload.version, payload.encoding, payload.resolving_conflict)


@router.patch('/{text_id}')
@errors
def metadata(text_id: int, payload: Meta):
    return texts.metadata(text_id, payload.favorite, payload.tags)


@router.post('/{text_id}/move')
@errors
def move(text_id: int, payload: Move):
    return texts.move(text_id, payload.directory, payload.name, payload.version)


@router.post('/{text_id}/trash')
@errors
def trash(text_id: int, payload: Version):
    return texts.trash(text_id, payload.version)


@router.get('/{text_id}/history')
@errors
def history(text_id: int):
    return texts.history(text_id)


@router.get('/{text_id}/history/{ident}')
@errors
def historical(text_id: int, ident: str):
    return texts.historical(text_id, ident)


@router.post('/{text_id}/history/{ident}/restore')
@errors
def restore(text_id: int, ident: str, payload: Version):
    h = texts.historical(text_id, ident)
    return texts.save(text_id, h['body'], payload.version, force_history=True)


@router.get('/{text_id}/asset')
@errors
def asset(text_id: int, href: str):
    path, _ = texts.asset(text_id, href)
    return FileResponse(path, headers={'X-Content-Type-Options': 'nosniff'})


@router.get('/{text_id}/asset-info')
@errors
def asset_info(text_id: int, href: str):
    target = texts.link_path(Path(texts.row(text_id)['path']), href)
    if target is not None and target.suffix.lower() in texts.EXTENSIONS:
        target = texts.document_path(target)
        linked = texts.conn().execute('SELECT id FROM texts WHERE path=? AND missing=0', (str(target),)).fetchone()
        if linked and target.is_file():
            return {'id': linked['id'], 'kind': 'text'}
    _, info = texts.asset(text_id, href)
    return info


@router.post('/{text_id}/reveal')
@errors
def reveal(text_id: int):
    import os
    import subprocess
    p = texts.document_path(texts.row(text_id)['path'])
    if not p.exists():
        raise HTTPException(404, '原文件不存在')
    if os.name == 'nt':
        subprocess.Popen(['explorer.exe', '/select,', str(p)])
    else:
        raise HTTPException(400, '请复制文件路径后在文件管理器中打开')
    return {'ok': True}


@router.get('/{text_id}/reference/{image_id}')
@errors
def reference(text_id: int, image_id: int):
    r = texts.conn().execute('SELECT path,kind,filename FROM images WHERE id=?', (image_id,)).fetchone()
    if not r:
        raise HTTPException(404, '素材不存在')
    href = texts.reference(Path(texts.row(text_id)['path']), Path(r['path']))
    label = r['filename'].replace('[', '').replace(']', '').replace('\n', '')
    return {'markdown': ('!' if r['kind'] == 'image' else '') + f'[{label}]({href})',
            'href': href, 'portable': not href.startswith('file:')}

File: frontend/src/lib/texts.ts
import { backendUrl } from './backend-url';

export interface TextSummary {
  id: number; path: string; filename: string; format: string; excerpt: string;
  mtime: number; created: number; opened: number; favorite: boolean; tags: string[];
  missing: boolean; error: string; match_offset?: number;
}
export interface TextDetail extends TextSummary {
  body: string; version: string; encoding: string | null; newline: string;
  bom: boolean; readonly: boolean; needs_encoding: boolean;
}
export interface TextHistory { id: string; text_id: number; created: number; filename: string; }
export class TextError extends Error { constructor(message: string, public status: number) { super(message); } }
async function request<T>(path: string, method = 'GET', payload?: unknown): Promise<T> {
  const response = await fetch(backendUrl('/api/texts' + path), {
    method, headers: { 'Content-Type': 'application/json' },
    ...(payload === undefined ? {} : { body: JSON.stringify(payload) }),
  });
  const result = await response.json();
  if (!response.ok) throw new TextError(typeof result.detail === 'string' ? result.detail : '文本操作失败', response.status);
  return result;
}
export const textsApi = {
  reorder: (ids: number[]) => request('/reorder', 'POST', { ids }),
  list: (options: Record<string, unknown> = {}) => {
    const q = new URLSearchParams();
    Object.entries(options).forEach(([k, v]) => { if (v !== null && v !== undefined && v !== '') q.set(k, String(v)); });
    return request<{ items: TextSummary[]; total: number }>('?' + q);
  },
  detail: (id: number, encoding?: string, opened = false) => request<TextDetail>(`/${id}?opened=${opened}${encoding ? '&encoding=' + encodeURIComponent(encoding) : ''}`),
  preview: (paths: string[]) => request<{ total: number; formats: Record<string, number>; roots: string[]; errors: { path: string; reason: string }[] }>('/preview', 'POST', { paths }),
  associate: (paths: string[]) => request<{ saved: TextSummary[]; errors: { path: string; reason: string }[] }>('/associate', 'POST', { paths }),
  create: (payload: { directory: string; name: string; body?: string; folder_id?: number | null }) => request<TextDetail>('', 'POST', payload),
  save: (id: number, body: string, version: string, encoding: string | null, resolving_conflict = false) => request<TextDetail>(`/${id}`, 'PUT', { body, version, encoding, resolving_conflict }),
  reveal: (id: number) => request(`/${id}/reveal`, 'POST'),
  meta: (id: number, payload: { favorite?: boolean; tags?: string[] }) => request<TextSummary>(`/${id}`, 'PATCH', payload),
  move: (id: number, directory: string, name: string, version: string) => request<TextDetail>(`/${id}/move`, 'POST', { directory, name, version }),
  trash: (id: number, version: string) => request(`/${id}/trash`, 'POST', { version }),
  history: (id: number) => request<TextHistory[]>(`/${id}/history`),
  historical: (id: number, entry: string) => request<{ body: string; created: number }>(`/${id}/history/${entry}`),
  restore: (id: number, entry: string, version: string) => request<TextDetail>(`/${id}/history/${entry}/restore`, 'POST', { version }),
  reference: (id: number, media: number) => request<{ markdown: string; href: string; portable: boolean }>(`/${id}/reference/${media}`),
  assetInfo: (id: number, href: string) => request<{ id: number; kind: 'image' | 'video' | 'text' }>(`/${id}/asset-info?href=${encodeURIComponent(href)}`),
};

export interface TextDraft { id: number; body: string; version: string; cursor: number; scroll: number; }
export type SaveState = 'saved' | 'dirty' | 'saving' | 'conflict' | 'error';

/** One save queue per document; responses never replace edits typed during a request. */
export class TextSession {
  body: string;
  baseline: string;
  state: SaveState = 'saved';
  error = '';
  composing = false;
  cursor = 0;
  scroll = 0;
  private timer?: ReturnType<typeof setTimeout>;
  private pending?: Promise<void>;
  private resolvingConflict = false;
  constructor(public detail: TextDetail, private changed: () => void,
    private api = textsApi, draft?: TextDraft) {
    this.body = this.baseline = detail.body;
    if (draft) {
      this.cursor = draft.cursor; this.scroll = draft.scroll;
      if (draft.body !== detail.body) {
        this.body = draft.body;
        this.state = draft.version === detail.version ? 'dirty' : 'conflict';
        if (this.state === 'conflict') this.error = '恢复了未保存草稿，磁盘文件也有改动';
      }
    }
  }
  get dirty() { return this.body !== this.baseline; }
  get draft(): TextDraft { return { id: this.detail.id, body: this.body, version: this.detail.version, cursor: this.cursor, scroll: this.scroll }; }
  edit(body: string) {
    this.body = body;
    if (!['conflict', 'error'].includes(this.state)) this.state = this.dirty ? 'dirty' : 'saved';
    this.changed(); this.schedule();
  }
  composition(active: boolean) { this.composing = active; if (active) clearTimeout(this.timer); else this.schedule(); }
  schedule() {
    clearTimeout(this.timer);
    if (this.dirty && !this.composing && !this.detail.readonly && !['conflict', 'error'].includes(this.state)) {
      this.timer = setTimeout(() => { void this.flush(); }, 1000);
    }
  }
  async flush(): Promise<void> {
    clearTimeout(this.timer);
    if (this.pending) { await this.pending; if (this.dirty && this.state === 'dirty') await this.flush(); return; }
    if (!this.dirty || this.composing || this.detail.readonly || this.state === 'conflict') return;
    const body = this.body;
    this.state = 'saving'; this.error = ''; this.changed();
    this.pending = (async () => {
      try {
        const result = await this.api.save(this.detail.id, body, this.detail.version, this.detail.encoding, this.resolvingConflict);
        this.resolvingConflict = false;
        this.detail = result; this.baseline = body;
        this.state = this.dirty ? 'dirty' : 'saved';
      } catch (error) {
        this.state = error instanceof TextError && error.status === 409 ? 'conflict' : 'error';
        this.error = error instanceof Error ? error.message : String(error);
      } finally { this.changed(); }
    })();
    await this.pending; this.pending = undefined; this.schedule();
  }
  async refresh() {
    if (this.pending || this.composing) return;
    const previous = this.detail.version;
    const result = await this.api.detail(this.detail.id);
    if (this.pending || this.detail.version !== previous) return;
    if (result.version !== previous) {
      if (this.dirty) { this.state = 'conflict'; this.error = '其他程序修改了原文件；自动保存已暂停'; }
      else { this.detail = result; this.body = this.baseline = result.body; this.state = 'saved'; }
    } else { this.detail = result; if (!this.dirty) { this.state = 'saved'; this.error = ''; } }
    this.changed();
  }
  acceptDisk(result: TextDetail) { this.detail = result; this.body = this.baseline = result.body; this.state = 'saved'; this.error = ''; this.changed(); }
  async keepMine() {
    // Deliberate conflict resolution, using a fresh version and still checking again on save.
    const disk = await this.api.detail(this.detail.id);
    this.detail = disk; this.baseline = disk.body; this.state = 'dirty';
    this.resolvingConflict = true;
    await this.flush();
  }
  dispose() { clearTimeout(this.timer); }
}
